.github/workflows/cicd.yml

```diff
@@ -0,0 +1,96 @@
+name: CICD
+on: [push]
+jobs:
+  Validate:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Clone Repo
+        uses: actions/checkout@v2
+        with:
+          persist-credentials: false
+
+      - name: Set Up NodeJs
+        uses: actions/setup-node@v2
+        with:
+          node-version: '14.17.4'
+          cache: npm
+
+      - name: Install Dependencies
+        run: npm ci
+
+      - name: Lint
+        run: npm run lint
+
+      - name: Test
+        run: npm test
+
+      - name: Build
+        run: npm run build
+
+      - name: Cache Deployment
+        if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/main' }}
+        uses: actions/cache@v2
+        env:
+          cache-name: deployment-cache
+        with:
+          path: './**/*'
+          key: ${{ github.event.after }}
+
+  Deploy:
+    needs: Validate
+    runs-on: ubuntu-latest
+    if: ${{ github.ref == 'refs/heads/main' }}
+    steps:
+      - name: Load Deployment
+        uses: actions/cache@v2
+        env:
+          cache-name: deployment-cache
+        with:
+          path: './**/*'
+          key: ${{ github.event.after }}
+
+      - name: Configure Git
+        env:
+          GH_EMAIL: ${{ secrets.GH_EMAIL }}
+          GH_TOKEN: ${{ secrets.GH_TOKEN }}
+          GH_USER: ${{ secrets.GH_USER }}
+        run: |
+          git config --local user.name "$GH_USER"
+          git config --local user.email "$GH_EMAIL"
+          git remote set-url origin "https://x-access-token:$GH_TOKEN@github.com/$GITHUB_REPOSITORY"
+
+      - name: Validate Git Connection
+        run: |
+          git ls-remote origin > /dev/null
+
+      - name: Create Git Release
+        run: |
+          npx standard-version --releaseCommitMessageFormat "chore(release): {{currentTag}} [skip ci]"
+
+      - name: Push Git Release
+        run: |
+          git push --follow-tags origin $GITHUB_REF
+
+      - name: Configure NPM
+        env:
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
+        run: |
+          npm set "//registry.npmjs.org/:_authToken=$NPM_TOKEN"
+
+      - name: Publish NPM Package
+        id: publish
+        run: |
+          npm publish
```

Comment:
The flow here is very intentionally made to cause the least disruption if something goes wrong. With our other product, we publish to npm and then to GitHub, but this can lead to GitHub failures leaving versions in `npm` which aren't accounted for in the repo's tags.

This flow instead will create the github release, then publish. If the publish fails, nothing broken or unaccounted for is shoved out into user-land (npm), and the Github release + changes are rolled back.